File: leetcode/16-3sum-closest.js
// https://leetcode.com/problems/3sum-closest/

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {

  nums.sort((a, b) => (a-b));
  let closestSum = 0;
  let minDiff = Number.MAX_SAFE_INTEGER;

  for (let i = 0; i < nums.length - 2; i++) {

    l = i + 1;
    r = nums.length - 1;

    while (l < r) {
      const sum = nums[i] + nums[l] + nums[r];

      // keep track of the closest sum to return it at the end...
      if (Math.abs(target - sum) < minDiff) {
        minDiff = Math.abs(target - sum);
        closestSum = sum;
      }

      // update the left and right pointers and if we found sum == target just return sum immediately
      if (sum < target) l += 1;
      else if (sum > target) r -= 1;
      else return sum;

    }
  }

  return closestSum;

};

File: leetcode/16-3sum-closest.py
# https://leetcode.com/problems/3sum-closest/

from typing import List

class Solution:
	def threeSumClosest(self, nums: List[int], target: int) -> int:
		closestSum = 0
		nums.sort()

		minDiff = float('inf')
		for i in range(len(nums) - 2):
			l = i+1
			r = len(nums) - 1
			while l < r:
				currSum = nums[i] + nums[l] + nums[r]

				if abs(target - currSum) < minDiff:
					closestSum = currSum
					minDiff = abs(target - currSum)

				if currSum < target:
					l += 1
				elif currSum > target:
					r -= 1
				else:
					return currSum

		return closestSum